core/src/main/java/com/linecorp/armeria/client/endpoint/RingHashEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint;
+
+import java.nio.charset.StandardCharsets;
+import java.util.SortedMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+import net.openhft.hashing.LongHashFunction;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+final class RingHashEndpointSelectionStrategy implements EndpointSelectionStrategy{
+
+    static final RingHashEndpointSelectionStrategy INSTANCE = new RingHashEndpointSelectionStrategy();
+    private RingHashEndpointSelectionStrategy() {}
+
+    @Override
+    public EndpointSelector newSelector(EndpointGroup endpointGroup) {
+        return new RingHashSelector(endpointGroup);
+    }
+    
+    static class RingHashSelector extends AbstractEndpointSelector {
+        private volatile WeightedRingEndpoint weightedRingEndpoint;
+
+        RingHashSelector(EndpointGroup endpointGroup) {
+            super(endpointGroup);
+            endpointGroup.addListener(endpoints -> 
+                weightedRingEndpoint = new WeightedRingEndpoint(endpoints), true
+            );
+        }
+
+        @Override
+        public @Nullable Endpoint selectNow(ClientRequestContext ctx) {
```

Comment:
We probably need `@Nullable` here because its super class has `@Nullable` already on this method? (We may want to change this, but at least for now)